IndexNode.cc

```diff
@@ -127,17 +138,40 @@ int IndexNode::getBlockAddress(int block)
  * less than the number of blocks in the file system
  * @exception java.lang.Exception if the block number is invalid
  */
-void IndexNode::setBlockAddress(int block , int address)
+void IndexNode::setBlockAddress(int block, int address) // Take in a filesystem structure?
 {
-	if(block >= 0 && block < MAX_DIRECT_BLOCKS)
-	{
+	FileSystem *fileSystem = Kernel::openFileSystems;
+	if(block >= 0 && block < MAX_DIRECT_BLOCKS){
 		directBlocks[block] = address ;
-	}
-	else
-	{
-		cout << "invalid block address " << block <<endl;
+		return;
 	}
 
+	if(block < fileSystem->getBlockSize() + MAX_DIRECT_BLOCKS) {
+	    int blockSize = fileSystem->getBlockSize();
+		int size = blockSize / sizeof(int); // blockSize should definitely be a multiple of 4, otherwise this might not work.
+		int ptrBlock[size];
+
+		if(indirectBlock == NOT_A_BLOCK){
+
+			// Pick an address for the indirectBlock
+			indirectBlock = fileSystem->allocateBlock();
+			if(indirectBlock < 0 ) {
+				return; // Maybe we should exit?
```

Comment:
What we really want to do is make `Kernel::write` or `read` return -1 in this case. To be thorough, the best solution is to change the function's signature to return an exit status. Then the caller checks for an error.

Reply:
That's hard tho

Reply:
Yeah that's true. Whoever wrote these functions originally just decided to skip the error checking here.